shared/js/ui/components/toggle.jsx

```diff
@@ -0,0 +1,203 @@
+         */
+        if (model.isBroken) {
+            if (!isBrowser()) {
+                toggleState.active = false
+                toggleState.disabled = true
+            }
+        }
+
+        const labelEnabled = ns.site('enableProtectionsSwitch.title')
+        const labelDisabled = ns.site('disableProtectionsSwitch.title')
+        toggleState.label = toggleState.active ? labelDisabled : labelEnabled
+
+        return toggleState
+    })
+
+    /**
+     * When the state changes, decide what to do
+     */
+    useEffect(() => {
+        if (!state.sideEffects) return
+
+        const int = setTimeout(() => {
+            // toggle, this communicates with native
+            model.toggleAllowlist()
+
+            // if the platform supports using a spinner
+            if (model.platformFeatures.spinnerFollowingProtectionsToggle) {
+                setState((prev) => {
+                    return { ...prev, toggled: true }
+                })
+            }
+        }, 300)
+        return () => {
+            clearTimeout(int)
+        }
+    }, [state.active, state.sideEffects])
+
+    /**
+     * Wrapper around the toggle method that consumers will call.
+     */
+    function toggle() {
+        setState((prev) => {
+            return { ...prev, active: false, sideEffects: true }
+        })
+    }
+
+    return [state, toggle]
+}
+
+/**
+ * @param {object} props
+ * @param {ToggleState} props.toggleState
+ * @param {() => void} props.onToggle
+ */
+export function ToggleButton(props) {
+    const { toggleState } = props
+
+    const labelEnabled = ns.site('enableProtectionsSwitch.title')
+    const labelDisabled = ns.site('disableProtectionsSwitch.title')
+    const label = toggleState.active ? labelDisabled : labelEnabled
+
+    // prettier-ignore
+    return isAndroid()
+        ? <AndroidToggle toggleState={toggleState} onToggle={props.onToggle} label={label} />
+        : <DefaultToggleButton toggleState={toggleState} label={label} onToggle={props.onToggle} />
+}
+
+/**
+ * @param {object} props
+ * @param {string} props.label
+ * @param {ToggleState} props.toggleState
+ * @param {() => void} props.onToggle
+ */
+export function AndroidToggle(props) {
+    const ref = useRef(null)
+    const className = `mdc-switch mdc-switch--${props.toggleState.active ? 'selected' : 'unselected'}`
+
+    useLayoutEffect(() => {
+        if (!ref.current) return
+        const elem = /** @type {HTMLButtonElement} */ (ref.current)
+        if (!(elem instanceof HTMLButtonElement)) return
+
+        const switchInstance = new MDCSwitch(ref.current)
+        switchInstance.listen('click', () => {
+            const pressed = elem.getAttribute('aria-checked')
+            const next = pressed === 'true' ? 'false' : 'true'
+            elem.setAttribute('aria-checked', next)
+            props.onToggle()
+            switchInstance.destroy()
+        })
+        return () => {
+            switchInstance.destroy()
+        }
+    }, [])
```

Comment:
This is only relevant to the Android toggle  - so there was no benefit to extracting this